### ISSUE TITLE
About Convolution Layer convert

### ISSUE THREAD

**Issue by wenching33** [daquexian/DNNConvertTool#6]
Hi 
When I converted a pretrained model, an error message showed "Only depthwise convolution or 
vanilla convolution are  supported". 
I  found the message at line 118 in caffe2daq.py and check my deploy.prototxt of that pretrained model.
It should be "group: 2" parameter setting of the convolution layer causing the converting failure.

My  question is : Is this also the limitation of NNAPI ? or User can additionally write code for this case and also other downward connections ?
Thanks in advance.

**Comment by daquexian** [comment]
Thanks for your using dnntools and DNNLibrary!


It is the limitation of NNAPI. Only standard conv and depthwise conv but
not group conv are supported.


I think you can try to apply group conv manually:
use ANEURALNETWORKS_STRIDED_SLICE(it's introduced in Android P so you need an Android P device or emulator to test it) to produce the two groups, apply conv on
them and concat two outputs. If you are not familiar with NNAPI I'm afraid
there is no straightforward way to implement it.

**Comment by daquexian** [comment]
I'm travelling now :D

If you are unfamiliar with NNAPI, you can wait for my update for group conv about 7 days later.

**Comment by daquexian** [comment]
BTW, what model are you trying to convert? :)

**Comment by wenching33** [comment]
Thanks. The pretrained model is a privately built and trained model similar to MTCNN.
I will check NNAPI first, Thanks for your replay.

**Comment by wenching33** [comment]
Hi 
I study caffe2daq.py and model_writer.py these days.
Is it possible to split convolution layer with group:2 (not depth-wise) during parsing the .caffemodel in caffe2daq.py (case: convolution), and call add_conv (which was defined in model_writer.py) twice.

I have also tried MMdnn to convert my .caffemodel to tensorflow model. I found the model_name.py produced by MMdnn involves code like:
---
def convolution(input, name, group, **kwargs):
    w = tf.Variable(__weights_dict[name]['weights'], trainable=is_train, name=name + "_weight")
    if group == 1:
        layer = tf.nn.convolution(input, w, **kwargs)
    else:
        weight_groups = tf.split(w, num_or_size_splits=group, axis=-1)
        xs = tf.split(input, num_or_size_splits=group, axis=-1)
        convolved = [tf.nn.convolution(x, weight, **kwargs) for
                    (x, weight) in zip(xs, weight_groups)]
        layer = tf.concat(convolved, axis=-1)
    if 'bias' in __weights_dict[name]:
        b = tf.Variable(__weights_dict[name]['bias'], trainable=is_train, name=name + "_bias")
        layer = layer + b
    return layer
---
It separates convolution layer with group usage into two vanilla conv. layer.
Is it possible to implement this kind of operation in conversion from .caffemodel to .daq ?
Thanks!

**Comment by daquexian** [comment]
@wenching33 Yes, it is possible. Since there is no "split" operator in NNAPI, and the only one operator that returns a sub-tensor of input is [strided slice](https://developer.android.com/ndk/reference/group/neural-networks#group___neural_networks_1ggaabbe492c60331b13038e39d4207940e0a89695302f8b1e7ae7ce8f4d8c0b8a752), you can use strided slice several times(one strided stride produces one group) to simulate "split" operator and get groups of input and conv weight.

Note: strided slice is introduced in Android P, so it need an Android P device or emulator.

**Comment by wenching33** [comment]
Hi ,
I'm trying to use strided slice operaiton. But I have to convert .caffemodel to .daq right?
Can we know how to seperation input from caffe model and save them to two vanilla convs layer in .daq format? 
Thanks

**Comment by daquexian** [comment]
Do you have qq, wechat or telegram? You can email me (daquexian566@gmail.com) your id so that we can chat directly on IM.

**Comment by wenching33** [comment]
Hi, I got weChat ID: luckier1981. Thanks a lot

**Comment by daquexian** [comment]
@wenching33 Group convolution has been supported in latest dnnlibrary and dnntools :D 

Feel free to reopen this issue if you still have a problem.
